File: js/hang/src/container/timeline.ts
/**
 * Publishing a media track's timeline: a companion track that maps each of the track's groups to
 * its start timestamp, so a consumer can seek (or build an HLS/DASH playlist) without downloading
 * the media. See the catalog {@link Catalog.Timeline} section that advertises it.
 *
 * @module
 */

import * as Json from "@moq/json";
import type * as Moq from "@moq/net";
import type { Time } from "@moq/net";
import type * as Catalog from "../catalog";
import { MOQ_EPOCH_UNIX_MILLIS, u53 } from "../catalog";

/** One timeline record: the media track opened `group` at presentation time `pts` (in the timeline's timescale). */
export interface Record {
	group: number;
	pts: number;
}

/** The default timeline timescale: 1000 units per second (milliseconds). */
export const DEFAULT_TIMESCALE = 1000;

/** The default record throttle: at most one record per second of media time. */
export const DEFAULT_GRANULARITY_MS = 1000;

/**
 * The conventional companion timeline track name for a media rendition: `<rendition>.timeline.z`
 * (the `.z` marks the DEFLATE-compressed stream, like the catalog's `.json.z` sibling).
 */
export function trackName(rendition: string): string {
	return `${rendition}.timeline.z`;
}

/** Options for a timeline {@link Producer}. */
export interface ProducerProps {
	/** Units per second for the records' `pts` (and the `wall` anchor). Defaults to milliseconds. */
	timescale?: number;

	/**
	 * Record at most one group per this much media time, in milliseconds. Video keyframes are
	 * already this far apart so every group is indexed; short audio groups are thinned out (a
	 * consumer extrapolates or fetches to fill a gap). Defaults to {@link DEFAULT_GRANULARITY_MS}.
	 */
	granularity?: number;
}

/**
 * Publishes one media track's timeline: an NDJSON record per group open, DEFLATE-compressed.
 *
 * {@link record} appends a group's start once. Advertise it in the rendition's catalog config via
 * {@link section}, and attach it to a {@link Legacy.Producer} (its `timeline` prop) to record group
 * opens automatically.
 */
export class Producer {
	#stream: Json.Stream.Producer<Record>;
	#track: string;
	#timescale: number;
	// The wall-clock time of pts 0, in timescale units since the moq epoch (advertised in the section).
	#wall?: number;
	// Minimum media-time gap between recorded groups (throttle), in microseconds.
	#granularityUs: number;
	// The pts (microseconds) of the last recorded group.
	#lastPts?: number;

	/** Wrap an already-created MoQ track (named per {@link trackName}) to publish a rendition's timeline. */
	constructor(track: Moq.Track, props: ProducerProps = {}) {
		this.#track = track.name;
		this.#timescale = props.timescale ?? DEFAULT_TIMESCALE;
		this.#granularityUs = (props.granularity ?? DEFAULT_GRANULARITY_MS) * 1000;
		this.#stream = new Json.Stream.Producer<Record>(track, { compression: true });
	}

	/** The catalog section advertising this timeline, to attach to the rendition's config. */
	section(): Catalog.Timeline {
		return {
			track: this.#track,
			timescale: u53(this.#timescale),
			wall: this.#wall === undefined ? undefined : u53(this.#wall),
		};
	}

	/**
	 * Set (or replace) the wall-clock anchor advertised in the catalog section, from an observed
	 * pairing of a media timestamp `pts` (microseconds) with its wall-clock time `wall` (defaulting
	 * to now). Stored as the extrapolated wall-clock time of pts 0, the single value the catalog
	 * `wall` field carries: in this timeline's timescale, measured from the moq epoch
	 * ({@link Catalog.MOQ_EPOCH_UNIX_MILLIS}, 2020). Throws if `wall` predates the moq epoch
	 * (unrepresentable).
	 */
	setWall(pts: Time.Micro, wall: Date = new Date()): void {
		const unixMillis = wall.getTime();
		if (unixMillis < MOQ_EPOCH_UNIX_MILLIS) {
			throw new Error(`wall time ${unixMillis} predates the moq epoch ${MOQ_EPOCH_UNIX_MILLIS}`);
		}
		const ptsUnits = Math.floor((pts * this.#timescale) / 1_000_000);
		const moqUnits = Math.floor(((unixMillis - MOQ_EPOCH_UNIX_MILLIS) * this.#timescale) / 1000);
		this.#wall = Math.max(0, moqUnits - ptsUnits);
	}

	/**
	 * Record that group `sequence` opened at presentation time `pts` (microseconds), unless it
	 * falls within the {@link ProducerProps.granularity} of the last recorded group (skipped, so a
	 * consumer extrapolates or fetches to fill the gap).
	 */
	record(sequence: number, pts: Time.Micro): void {
		if (this.#lastPts !== undefined && pts < this.#lastPts + this.#granularityUs) return;
		this.#lastPts = pts;
		this.#stream.append({ group: sequence, pts: Math.floor((pts * this.#timescale) / 1_000_000) });
	}

	/** Finish the timeline track. */
	finish(): void {
		this.#stream.finish();
	}
}

File: rs/hang/src/timeline.rs
//! The timeline track: an ordered log mapping each of a media track's groups to its start
//! timestamp.
//!
//! MoQ groups carry only an opaque sequence number; the timestamps live inside the media
//! frames. A timeline track republishes that mapping as metadata, so a consumer can answer
//! "which group covers time T" (and "where is the live edge") from a few bytes per group
//! instead of downloading the media itself. That is the primitive an HLS/DASH origin needs to
//! render playlists without touching media bytes, and the index a VOD player seeks with.
//!
//! A timeline is per media track (audio and video groups have different durations, so they
//! can't share one). The media track's catalog entry carries a
//! [`Timeline`](crate::catalog::Timeline) section naming the companion timeline track and its
//! timescale (`pts` is in those units, default milliseconds) and optional wall-clock anchor.
//! This module is just the wire record on that track.
//!
//! On the wire the track is a `moq-json` *stream* (see `moq_json::stream`): a single group, one
//! DEFLATE-compressed record per frame. The record shape follows the `mediatimeline` concept
//! from [draft-ietf-moq-msf](https://datatracker.ietf.org/doc/draft-ietf-moq-msf/), with JSON
//! keys instead of positional arrays (DEFLATE absorbs the repeated keys) and a moq-lite group
//! sequence instead of a `[group, object]` location (moq-lite has no object IDs).
//!
//! Like the catalog, a record tolerates and preserves unknown fields: extend it by flattening
//! a `Record` into your own struct, or read its `ext` field directly.

use serde::{Deserialize, Serialize};

use crate::Result;

/// The application extension carried alongside a record's `group`/`pts`.
///
/// Defaults to `()` (no extra fields). Set an application's own typed struct to add fields
/// (e.g. a discontinuity flag, a measured bitrate); it is flattened into the record's JSON
/// object, exactly like [`Catalog`](crate::Catalog)'s extension. `()` is the base case.
pub trait RecordExt: serde::Serialize + serde::de::DeserializeOwned + Default + Clone + Send + Unpin + 'static {}
impl RecordExt for () {}

/// One timeline record: the media track opened group `group` at presentation time `pts`.
///
/// Records are appended when the media group opens, so the live edge of the timeline is the
/// live edge of the broadcast. A group's duration is implicit: the gap to the next record.
/// `pts` is in the timescale declared by the track's [`Timeline`](crate::catalog::Timeline)
/// catalog section (default milliseconds). Extend it with a typed [`RecordExt`].
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(
	rename_all = "camelCase",
	bound(serialize = "E: serde::Serialize", deserialize = "E: serde::de::DeserializeOwned")
)]
pub struct Record<E: RecordExt = ()> {
	/// The group's sequence number, as used by FETCH/SUBSCRIBE on the media track.
	pub group: u64,

	/// The group's start (its first frame's presentation timestamp), in the timeline's
	/// timescale.
	pub pts: u64,

	/// The application extension, flattened into the record's JSON object (nothing for the
	/// default `()`). See [`RecordExt`].
	#[serde(flatten)]
	pub ext: E,
}

impl<E: RecordExt> Record<E> {
	/// A record with the default (empty) extension.
	pub fn new(group: u64, pts: u64) -> Self {
		Self {
			group,
			pts,
			ext: E::default(),
		}
	}

	/// Parse a record from a slice of bytes.
	pub fn from_slice(v: &[u8]) -> Result<Self> {
		Ok(serde_json::from_slice(v)?)
	}

	/// Serialize the record to a vector of bytes.
	pub fn to_vec(&self) -> Result<Vec<u8>> {
		Ok(serde_json::to_vec(self)?)
	}
}

/// The conventional companion timeline track name for a media rendition: `<rendition>.timeline.z`
/// (the `.z` marks the DEFLATE-compressed stream, like the catalog's `.json.z` sibling).
///
/// A publisher names the timeline track this way and records it in the media track's
/// [`Timeline::track`](crate::catalog::Timeline::track) catalog field; a consumer reads the
/// name from the catalog rather than reconstructing it, so this is only a default.
pub fn track_name(rendition: &str) -> String {
	format!("{rendition}.timeline.z")
}

#[cfg(test)]
mod test {
	use super::*;

	#[test]
	fn roundtrip() {
		let record = Record::<()>::new(42, 84_000);
		let json = record.to_vec().unwrap();
		assert_eq!(std::str::from_utf8(&json).unwrap(), r#"{"group":42,"pts":84000}"#);
		assert_eq!(Record::<()>::from_slice(&json).unwrap(), record);
	}

	#[test]
	fn typed_extension_flattens() {
		// An application extends the record with its own typed section, flattened into the object.
		#[derive(serde::Serialize, serde::Deserialize, Default, Clone, PartialEq, Debug)]
		struct Ext {
			#[serde(skip_serializing_if = "std::ops::Not::not", default)]
			discontinuity: bool,
		}
		impl RecordExt for Ext {}

		let record = Record {
			group: 7,
			pts: 14_000,
			ext: Ext { discontinuity: true },
		};
		let json = record.to_vec().unwrap();
		assert_eq!(
			std::str::from_utf8(&json).unwrap(),
			r#"{"group":7,"pts":14000,"discontinuity":true}"#
		);
		assert_eq!(Record::<Ext>::from_slice(&json).unwrap(), record);
	}

	#[test]
	fn conventional_track_name() {
		assert_eq!(track_name("video0"), "video0.timeline.z");
	}
}
